src/Classes/TradeQuery.lua

```diff
@@ -0,0 +1,897 @@
+		end
+		if validURL then
+			activeSlotRef.pbURL = controls['uri'..str_cnt].buf
+		elseif controls['uri'..str_cnt].buf == "" then
+			activeSlotRef.pbURL = ""
+		end
+		return validURL and self:PriceBuilderCanSearch(controls) and self:PriceBuilderCanFetch(controls)
+	end
+	controls['resultIndex'..str_cnt] = new("EditControl", {"TOPLEFT",controls['priceButton'..str_cnt],"TOPLEFT"}, 100 + 8, 0, 60, row_height, "#", nil, "%D", 3, function(buf)
+		controls['resultIndex'..str_cnt].buf = tostring(m_min(m_max(tonumber(buf) or 1, 1), self.sortedResultTbl[str_cnt] and #self.sortedResultTbl[str_cnt] or 1))
+	end)
+	controls['resultIndex'..str_cnt].tooltipFunc = function(tooltip)
+		if tooltip:CheckForUpdate(controls['resultIndex'..str_cnt].buf) then
+			self:SetFetchResultReturn(controls, str_cnt, tonumber(controls['resultIndex'..str_cnt].buf))
+		end
+	end
+	controls['resultCount'..str_cnt] = new("EditControl", {"TOPLEFT",controls['resultIndex'..str_cnt],"TOPLEFT"}, 60 + 8, 0, 82, row_height, "^8No Results", nil, nil)
+	controls['priceAmount'..str_cnt] = new("EditControl", {"TOPLEFT",controls['resultCount'..str_cnt],"TOPLEFT"}, 82 + 8, 0, 120, row_height, "Price", nil, nil)
+	controls['priceAmount'..str_cnt].enabled = function()
+		local boolean = #controls['priceAmount'..str_cnt].buf > 0
+		return boolean
+	end
+	controls['importButton'..str_cnt] = new("ButtonControl", {"TOPLEFT",controls['priceAmount'..str_cnt],"TOPLEFT"}, 120 + 8, 0, 100, row_height, "Import Item", function()
+		self.itemsTab:CreateDisplayItemFromRaw(controls['importButtonText'..str_cnt].buf)
+		local item = self.itemsTab.displayItem
+		-- pass "true" to not auto equip it as we will have our own logic
+		self.itemsTab:AddDisplayItem(true)
+		-- Autoequip it
+		local slot = slotTbl.ref and self.itemsTab.sockets[slotTbl.ref] or self.itemsTab.slots[slotTbl.name]
+		if slot and slotTbl.name == slot.label and slot:IsShown() and self.itemsTab:IsItemValidForSlot(item, slot.slotName) then
+			slot:SetSelItemId(item.id)
+			self.itemsTab:PopulateSlots()
+			self.itemsTab:AddUndoState()
+			self.itemsTab.build.buildFlag = true
+		end
+	end)
+	controls['importButton'..str_cnt].tooltipFunc = function(tooltip)
+		tooltip:Clear()
+		if #controls['importButtonText'..str_cnt].buf > 0 then
+			local item = new("Item", controls['importButtonText'..str_cnt].buf)
+			self.itemsTab:AddItemTooltip(tooltip, item, nil, true)
+		end
+	end
+	controls['importButton'..str_cnt].enabled = function()
+		return #controls['importButtonText'..str_cnt].buf > 0
+	end
+	-- for storing the base64 item description
+	controls['importButtonText'..str_cnt] = new("EditControl", nil, 0, 0, 0, 0, "", nil, "", nil, nil, 16)
+	controls['importButtonText'..str_cnt].shown = false
+	-- for storing the whisper to purchase item from seller
+	controls['whisperButtonText'..str_cnt] = new("EditControl", nil, 0, 0, 0, 0, "", nil, "", nil, nil, 16)
+	controls['whisperButtonText'..str_cnt].shown = false
```

Comment:
Why are we storing text in hidden UI controls?  Seems like a terrible code smell and there has to be a better data structure that can hold them.  The results end up getting wiped if you close the modal also, which means all of the slots have to be searched again.

Reply:
Yeah... I'm torn on that one. Due to the subscripts running in their own thread and my unfamiliarity with how they interface with shared memory made me re-use what I knew worked (aka `controls`). You need a structure that can be access from another thread and asynchronously updated. 

Welcome other approaches

Reply:
Okay, I'll take a look once I dive into the code and see if there's another way.  Coroutines definitely make it more complex than when I first commented on it, but UI elements as data storage is pretty icky even still.